### PR TITLE
Conv1DLayer/Dense tests, LogSoftmax testing, all for ResNet_1D validation

### PR DESCRIPTION
**Pull request by mcmillan03**
Everything needed to support a ResNet1D model… added support for 1,2,3 input channels.  Adding support of "odd channels" by introducing a 5th shape term (logical channels) and adding channel padding where necessary. New intrinsics, functions and layer classes for SoftMax and LogSoftMax and Dense.  Adding some missing 1D layer classes and minimal test code. Added macro for compile time ability to swap front/back padding values (e.g. between pytorch and TF padding).  Lots of miscellaneous code cleanup and making more consistent.

### CODE REVIEW

File: include/small/AveragePool2DLayer.hpp
Comment (UpasanaSridhar) [inline include/small/AveragePool2DLayer.hpp:1]: It is a little confusing to set output shape with the following order:  this->set_output_shape(
            {m_input_shape[BATCH],
             m_input_shape[CHANNEL],
             H, W,
             m_input_shape[L_CHAN]});

and then use it in an oder where L_CHAN and CHANNEL are used together. 

Not a major gripe. I think the logical channels should be separated into a separate thing entirely, with a logical shape (view) and a data layout (current shape)

File: include/small/Layer.hpp [hunk 1]
```diff
@@ -20,10 +20,16 @@
...
 {
-
-// Currently only for Conv2DLayer
+/**
+ * "In-place activation functions that can used in DenseLayer,
+ * Conv2D/Conv1DLayer, DepthwiseConv2DLayer, and PartialConv2DLayer
+ *
+ * @note LogSoftMax CANNOT be added here (it's not in-place)
+ *
+ * @todo Consider removing activations from within other layer classes
```
Comment (UpasanaSridhar) [inline include/small/Layer.hpp:29]: I don't understand this comment. Remove ReLU from ConvLayer.cpp?

File: include/small/Layer.hpp [hunk 2]
```diff
@@ -37,16 +43,23 @@ class Layer
...
 
     inline size_t output_size() const
     {
-        return m_output_size;
+        return m_output_shape.size();
+    }
+
+    inline size_t logical_output_size() const
+    {
+        return m_output_shape.logical_size();
```
Comment (UpasanaSridhar) [inline include/small/Layer.hpp:57]: What is the difference between m_output_shape and output_shape/input_shape

**Review by UpasanaSridhar — APPROVED** [review]
We should really have a way to instantiate all the different incarnations of a layer more systematically. It is hard to review changes in such a large file when most things remain the same.

File: include/small/AveragePool2DLayer.hpp
Comment (mcmillan03) [reply include/small/AveragePool2DLayer.hpp:1]: Deferring this.  Will deal with Tensor manipulation ops in a different branch.

File: include/small/Layer.hpp (hunk 1 above)
Comment (mcmillan03) [reply include/small/Layer.hpp:29]: I want to eventually separate the activations from all of the Conv layers.  Deferring for now.
